test/gtest/tl/mlx5/test_tl_mlx5_wqe.cc

```diff
@@ -0,0 +1,317 @@
+
+    *src = 0xdeadbeef;
+    //memory barrier
+    ucc_memory_cpu_fence();
+    *ctrl = 1;
+
+    while (1) {
+        completions_num = ibv_poll_cq(cq, 1, wcs);
+        if (completions_num != 0) {
+            GTEST_ASSERT_EQ(completions_num, 1);
+            GTEST_ASSERT_EQ(wcs[0].status, IBV_WC_SUCCESS);
+            if (wcs[0].wr_id == 0) {
+                break;
+            }
+        }
+    }
+
+    //validation
+    EXPECT_EQ(*dst, *src);
+
+    GTEST_ASSERT_EQ(ibv_dereg_mr(buffer_mr), UCC_OK);
+}
+
+UCC_TEST_P(test_tl_mlx5_umr_wqe, umrWqe)
+{
+    uint16_t nbr_srcs     = std::get<0>(GetParam());
+    uint32_t bytes_count  = std::get<1>(GetParam());
+    uint32_t repeat_count = std::get<2>(GetParam());
+    uint32_t bytes_skip   = std::get<3>(GetParam());
+    int      src_size = (bytes_count + bytes_skip) / sizeof(DT) * repeat_count;
+    int      dst_size = bytes_count / sizeof(DT) * nbr_srcs * repeat_count;
+    int      send_mem_access_flags = 0;
+    int      recv_mem_access_flags =
+        IBV_ACCESS_LOCAL_WRITE | IBV_ACCESS_REMOTE_WRITE;
+    int                          completions_num = 0;
+    DT                           src[nbr_srcs][src_size], dst[dst_size];
+    struct ibv_mr *              src_mr[nbr_srcs], *dst_mr, *umr_entries_mr;
+    struct mlx5dv_mkey *         umr_mkey;
+    struct mlx5dv_mkey_init_attr umr_mkey_init_attr;
+    void *                       umr_entries_buf;
+    size_t                       umr_buf_size;
+    struct mlx5dv_mr_interleaved mkey_entries[nbr_srcs];
+    struct ibv_wc                wcs[1];
+    int                          i, j, src_index, repetition_index, offset;
+
+    // We get an error when the number of src is a multiple of 4.
+    // In this case the UMR WQE returns with a status IBV_WC_MW_BIND_ERR
+    if (nbr_srcs % 4 == 0)
+    {
+        GTEST_SKIP();
+    }
+
+    // Setup src and dst buffers
+    for (i = 0; i < nbr_srcs; i++) {
+        for (j = 0; j < src_size; j++) {
+            src[i][j] = (i * src_size + j) % 256;
+        }
+        src_mr[i] = ibv_reg_mr(pd, src[i], src_size, send_mem_access_flags);
+        GTEST_ASSERT_NE(nullptr, src_mr[i]);
+    }
+
+    memset(dst, 0, dst_size);
+    dst_mr = ibv_reg_mr(pd, dst, dst_size, recv_mem_access_flags);
+    GTEST_ASSERT_NE(nullptr, dst_mr);
+
+    // UMR
+    umr_buf_size = ucc_align_up(
+        sizeof(struct mlx5_wqe_umr_repeat_ent_seg) * (nbr_srcs + 1), 64);
+    GTEST_ASSERT_EQ(ucc_posix_memalign(&umr_entries_buf, 2048, umr_buf_size),
+                    0);
+
+    umr_entries_mr =
+        ibv_reg_mr(pd, umr_entries_buf, umr_buf_size, send_mem_access_flags);
+    GTEST_ASSERT_NE(nullptr, umr_entries_mr);
+
+    memset(&umr_mkey_init_attr, 0, sizeof(umr_mkey_init_attr));
+    umr_mkey_init_attr.pd           = pd;
+    umr_mkey_init_attr.create_flags = MLX5DV_MKEY_INIT_ATTR_FLAGS_INDIRECT;
+    umr_mkey_init_attr.max_entries  = nbr_srcs;
+
+    umr_mkey = mlx5dv_create_mkey(&umr_mkey_init_attr);
```

Comment:
check max_entries is greater or equal than nbr_srcs on function return

Reply:
I am not sure to understand.
I added a check that the qp's `max_inline_data` is greater than `nbr_srcs`. Does it address your comment ?